util/linuxfw/nftables_runner.go

```diff
@@ -0,0 +1,1064 @@
+				Register: 1,
+				Data:     netip,
+			},
+			&expr.Counter{},
+			&expr.Verdict{
+				Kind: expr.VerdictReturn,
+			},
+		},
+	}
+	return rule, nil
+}
+
+// addReturnChromeOSVMRangeRule adds a rule to return if the source IP
+// is in the ChromeOS VM range.
+func addReturnChromeOSVMRangeRule(c *nftables.Conn, table *nftables.Table, chain *nftables.Chain, tunname string) error {
+	rule, err := createReturnChromeOSVMRangeRule(table, chain, tunname)
+	if err != nil {
+		return fmt.Errorf("create rule: %w", err)
+	}
+	_ = c.AddRule(rule)
+	if err = c.Flush(); err != nil {
+		return fmt.Errorf("add rule: %w", err)
+	}
+	return nil
+}
+
+// createDropCGNATRangeRule creates a rule to drop if the source IP is
+// in the CGNAT range, and the interface is not the tunname.
+func createDropCGNATRangeRule(table *nftables.Table, chain *nftables.Chain, tunname string) (*nftables.Rule, error) {
+	saddrExpr, err := newLoadSaddrExpr(nftables.TableFamilyIPv4)
+	if err != nil {
+		return nil, fmt.Errorf("get expr: %w", err)
+	}
+	_, ipNet, err := net.ParseCIDR(tsaddr.CGNATRange().String())
+	if err != nil {
+		return nil, fmt.Errorf("parse cidr: %w", err)
+	}
+	mask, err := hex.DecodeString(ipNet.Mask.String())
+	if err != nil {
+		return nil, fmt.Errorf("decode mask: %w", err)
+	}
+	netip := ipNet.IP.Mask(ipNet.Mask).To4()
+	rule := &nftables.Rule{
+		Table: table,
+		Chain: chain,
+		Exprs: []expr.Any{
+			&expr.Meta{Key: expr.MetaKeyIIFNAME, Register: 1},
+			&expr.Cmp{
+				Op:       expr.CmpOpNeq,
+				Register: 1,
+				Data:     []byte(tunname),
+			},
+			saddrExpr,
+			&expr.Bitwise{
+				SourceRegister: 1,
+				DestRegister:   1,
+				Len:            4,
+				Mask:           mask,
+				Xor:            []byte{0x00, 0x00, 0x00, 0x00},
+			},
+			&expr.Cmp{
+				Op:       expr.CmpOpEq,
+				Register: 1,
+				Data:     netip,
+			},
+			&expr.Counter{},
+			&expr.Verdict{
+				Kind: expr.VerdictDrop,
+			},
+		},
+	}
+	return rule, nil
+}
+
+// addDropCGNATRangeRule adds a rule to drop if the source IP is in the
+// CGNAT range.
+func addDropCGNATRangeRule(c *nftables.Conn, table *nftables.Table, chain *nftables.Chain, tunname string) error {
+	rule, err := createDropCGNATRangeRule(table, chain, tunname)
+	if err != nil {
+		return fmt.Errorf("create rule: %w", err)
+	}
+	_ = c.AddRule(rule)
+	if err = c.Flush(); err != nil {
+		return fmt.Errorf("add rule: %w", err)
+	}
+	return nil
+}
+
+// createSetSubnetRouteMarkRule creates a rule to set the subnet route
+// mark if the packet is from the given interface.
+func createSetSubnetRouteMarkRule(table *nftables.Table, chain *nftables.Chain, tunname string) (*nftables.Rule, error) {
+	hexTsFwmarkMaskNeg, err := decodeHexString(4, TailscaleFwmarkMaskNeg)
+	if err != nil {
+		return nil, fmt.Errorf("decode hex string: %w", err)
+	}
+	hexTSSubnetRouteMark, err := decodeHexString(4, TailscaleSubnetRouteMarkHexStr)
+	if err != nil {
+		return nil, fmt.Errorf("decode hex string: %w", err)
+	}
```

Comment:
Yeah, I removed the string part and just used Decode. I still want to encapsulate this logic in a function to avoid a change by mistake. I didn't use PutUint32 tho, I just used Hex.Decode like 
```
	ret := make([]byte, toLen)
	hex.Decode(ret, []byte(hexStr))
```
These I believe should output the same result. 

Regarding that marks are big endian, we can find that the nftables package tested their function in big endian.

Reply:
You know what, I'm still overthinking this. `binary.BigEndian.PutUint32` is still pretty annoying to use. But it's silly to parse a hex string each time. It's a 32-bit constant that needs to be passed to the kernel via netlink as big endian.

Maybe try something like:
```golang
func getSubnetRouteMark() [4]byte {
    return [4]byte{0x00, 0x04, 0x00, 0x00}
}
```

That avoids all fiddling around with temporary buffers, allocations, or errors that need to be handled.

Reply:
Switched to function